# Lagrangian optimization example

$\renewcommand{\vec}[1]{\mathbf{#1}} \newcommand{\mat}[1]{\mathtt{#1}}$
When we want to find a minimum or maximum of a function $f(\vec{w})$ subject to a constraint $g(\vec{w}) = 0$,
we can use the Lagrangian approach to solving it.

$\newcommand{\lgr}{{\cal L}}$
The Lagrangian is the function $\lgr(\vec{w},\lambda) = f(\vec{w}) + \lambda g(\vec{w})$.

If we can find $(\vec{w}^*,\lambda^*)$, a stationary point of $\lgr$, then $\vec{w}^*$ is a stationary point of $f$.

In class we had the simple example $f(\vec{w}) = \|w\|^2$ and $g(\vec{w}) = w_1 + w_2 - 4$.

By inspection, we can easily see that the solution is $w_1 = w_2 = 2$.

For the analysis, we write the Lagrangian $\lgr(\vec{w},\lambda) = w_1^2 + w_2^2 + \lambda(w_1 + w_2 - 4)$.

Clearly, $$\frac{\partial \lgr}{\partial w_1} = 2 w_1 + \lambda$$
$$\frac{\partial \lgr}{\partial w_2} = 2 w_2 + \lambda$$
$$\frac{\partial \lgr}{\partial \lambda} = w_1 + w_2 - 4.$$

To obtain the stationariy point(s) of this system, we set the partials to 0 and solve. From
$$\lambda = -2w_1 = -2w_2,$$
we obtain $w_1 = w_2$ then the solution $w_1^* = 2, w_2^* = 2, \lambda^* = -4$.

We already knew $(w_1,w_2) = (2,2)$ by inspection, but now we have confirmed it analytically. We also know it's a minimum by inspection, but to confirm this analytically, we need the Hessian
$$\mat{H}_\lgr = \begin{bmatrix}
\frac{\partial^2 \lgr}{\partial^2 \lambda} & \frac{\partial^2 \lgr}{\partial \lambda \partial w_1} & \frac{\partial^2 \lgr}{\partial \lambda \partial w_2} \\
\frac{\partial^2 \lgr}{\partial w_1 \partial \lambda} & \frac{\partial^2 \lgr}{\partial^2 w_1} & \frac{\partial^2 \lgr}{\partial w_1 \partial w_2} \\ \frac{\partial^2 \lgr}{\partial w_2 \partial \lambda} & \frac{\partial^2 \lgr}{\partial w_2 \partial w_1} & \frac{\partial^2 \lgr}{\partial^2 w_2} \end{bmatrix} = 
\begin{bmatrix} 0 & 1 & 1 \\ 1 & 2 & 0 \\ 1 & 0 & 2 \end{bmatrix} $$
Take a look at the eigenvalues of $\mat{H}_\lgr$:

In [4]:
import numpy as np

Hl = np.array([[0, 1, 1], [1, 2, 0], [1, 0, 2]])
eigvals, eigvecs = np.linalg.eig(Hl)
print("Eigenvalues: %f, %f, %f" % (eigvals[0], eigvals[1], eigvals[2]))

Eigenvalues: -0.732051, 2.732051, 2.000000


Since we have a mixture of positive and negative eigenvalues (-0.732051, 2.732051, and 2.0), the critical point we've found is a *saddle point* of the Lagrangian. It is actually always the case that the solution of the original optimization is a saddle point of the Lagrangian function.

To determine if the point is a minimum, maximum, or saddle point of the original optimization problem, we consider the determinant of various submatrices of $\mat{H}_\lgr$.

For $m$ constraints, we consider the $2m+1$-th up to the $n$th principal minors of $\mat{H}_\lgr$. These are the upper left square submatrices of $\mat{H}_\lgr$. In our case, $m = 1 $ and $n = 3 = 2m+1$, so we have only one principal minor to consider, $\mat{H}_\lgr$ itself. We get its determinant:

In [6]:
print("Determinant of Hl: %f" % np.linalg.det(Hl))

Determinant of Hl: -4.000000


For a minimum, all of the minors' determinants must have a sign of $(-1)^m$ which means negative for $m=1$. Since this is true, we can conclude that the critical point is a minimum of $f(\vec{w})$! 